File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #Usual Prefix , Postfix 
        #Notice that we write res[i] before prefix or postfix calculation ,this is not to include value at nums[i]
        res = [0] * len(nums)

        prefix = 1 

        for i in range(len(nums)):
            res[i] = prefix 
            prefix *= nums[i]

        postfix = 1 

        for i in range(len(nums) - 1 , -1 , -1) : 
            res[i] *= postfix 
            postfix *= nums[i]

        return res 

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];

        res[0] = 1;
        for(int i = 1; i < n;i ++){
            res[i] = res[i - 1] * nums[i - 1];
        }

        int postfix = 1;
        for(int i = n - 1; i >= 0; i --){
            res[i] *= postfix;
            postfix *= nums[i];
        }

        return res;
    }
}